scrapy/core/engine.py

```diff
@@ -54,269 +64,277 @@ def _maybe_fire_closing(self):
-                f"{type(response)}: {response!r}"
-            )
+    def _handle_downloader_output(
+        self, result: Union[Request, Response, Failure], request: Request
+    ) -> Optional[Deferred]:
+        assert self.spider is not None  # typing
+
+        if not isinstance(result, (Request, Response, Failure)):
+            raise TypeError(f"Incorrect type: expected Request, Response or Failure, got {type(result)}: {result!r}")
+
         # downloader middleware can return requests (for example, redirects)
-        if isinstance(response, Request):
-            self.crawl(response, spider)
-            return
-        # response is a Response or Failure
-        d = self.scraper.enqueue_scrape(response, request, spider)
-        d.addErrback(lambda f: logger.error('Error while enqueuing downloader output',
-                                            exc_info=failure_to_exc_info(f),
-                                            extra={'spider': spider}))
+        if isinstance(result, Request):
+            self.crawl(result)
+            return None
+
+        d = self.scraper.enqueue_scrape(result, request, self.spider)
+        d.addErrback(
+            lambda f: logger.error(
+                "Error while enqueuing downloader output",
+                exc_info=failure_to_exc_info(f),
+                extra={'spider': self.spider},
+            )
+        )
         return d
 
-    def spider_is_idle(self, spider):
+    def spider_is_idle(self, spider: Optional[Spider] = None) -> bool:
+        if spider is not None:
+            warnings.warn(
+                "Passing a 'spider' argument to ExecutionEngine.spider_is_idle is deprecated",
+                category=ScrapyDeprecationWarning,
+                stacklevel=2,
+            )
+        if self.slot is None:
+            raise RuntimeError("Engine slot not assigned")
         if not self.scraper.slot.is_idle():
-            # scraper is not idle
             return False
-
-        if self.downloader.active:
-            # downloader has pending requests
+        if self.downloader.active:  # downloader has pending requests
             return False
-
-        if self.slot.start_requests is not None:
-            # not all start requests are handled
+        if self.slot.start_requests is not None:  # not all start requests are handled
             return False
-
         if self.slot.scheduler.has_pending_requests():
-            # scheduler has pending requests
             return False
-
         return True
 
-    @property
-    def open_spiders(self):
-        return [self.spider] if self.spider else []
-
-    def has_capacity(self):
-        """Does the engine have capacity to handle more spiders"""
-        return not bool(self.slot)
-
-    def crawl(self, request, spider):
-        if spider not in self.open_spiders:
-            raise RuntimeError(f"Spider {spider.name!r} not opened when crawling: {request}")
```

Comment:
Shouldn’t this error still be caught, if the spider argument does not match `self.spider`? (maybe with a better error message that explicitly says so, instead of the old message)

Reply:
Good call. It's updated, hopefully this is more explicit:
```python
def crawl(self, request: Request, spider: Optional[Spider] = None) -> None:
    """Inject the request into the spider <-> downloader pipeline"""
    if spider is not None:
        warnings.warn(
            "Passing a 'spider' argument to ExecutionEngine.crawl is deprecated",
            category=ScrapyDeprecationWarning,
            stacklevel=2,
        )
        if spider is not self.spider:
            raise RuntimeError(f"The spider '{spider.name!r}' does not match the open spider")
    if self.spider is None:
        raise RuntimeError(f"No open spider to crawl: {request}")
    self._schedule_request(request, self.spider)
    self.slot.nextcall.schedule()  # type: ignore[union-attr]
```

Note that I added the same verification code at `download`, which at the moment does not check if the passed spider matches the open one. This is technically backward-incompatible, but I think it's a reasonable thing to do.

Reply:
> Note that I added the same verification code at download, which at the moment does not check if the passed spider matches the open one. This is technically backward-incompatible, but I think it's a reasonable thing to do.

Why not just log an additional warning instead, or even an error? It should still catch the attention of anyone actually doing this (hopefully no one), while still remaining backward compatible.